src/main/scala/analysis/UtilMethods.scala

```diff
@@ -54,3 +56,85 @@ def evaluateExpression(exp: Expr, constantPropResult: Map[Variable, FlatElement[
       None
   }
 }
+
+def evaluateExpressionWithSSA(exp: Expr, constantPropResult: Map[RegisterVariableWrapper, Set[BitVecLiteral]]): Set[BitVecLiteral] = {
+  Logger.info(s"evaluateExpression: $exp")
+
+  def apply(op: (BitVecLiteral, BitVecLiteral) => BitVecLiteral, a: Set[BitVecLiteral], b: Set[BitVecLiteral]): Set[BitVecLiteral] =
+    val res = for {
+      x <- a
+      y <- b
+    } yield op(x, y)
+    res
+
+  def applySingle(op: BitVecLiteral => BitVecLiteral, a: Set[BitVecLiteral]): Set[BitVecLiteral] =
+    val res = for {
+      x <- a
+    } yield op(x)
+    res
+
+
+  exp match {
+    case binOp: BinaryExpr =>
+      val lhs = evaluateExpressionWithSSA(binOp.arg1, constantPropResult)
+      val rhs = evaluateExpressionWithSSA(binOp.arg2, constantPropResult)
+
+      (lhs, rhs) match {
+        case (l: Set[BitVecLiteral], r: Set[BitVecLiteral]) =>
+          binOp.op match {
+            case BVADD => apply(BitVectorEval.smt_bvadd, lhs, rhs)
+            case BVSUB => apply(BitVectorEval.smt_bvsub, lhs, rhs)
+            case BVASHR => apply(BitVectorEval.smt_bvashr, lhs, rhs)
+            case BVCOMP => apply(BitVectorEval.smt_bvcomp, lhs, rhs)
+            case BVCONCAT => apply(BitVectorEval.smt_concat, lhs, rhs)
+            case BVLSHR => apply(BitVectorEval.smt_bvlshr, lhs, rhs)
+            case BVSHL => apply(BitVectorEval.smt_bvshl, lhs, rhs)
+            case BVOR => apply(BitVectorEval.smt_bvor, lhs, rhs)
+            case _ => throw new RuntimeException("Binary operation support not implemented: " + binOp.op)
+          }
+      }
+    case extend: ZeroExtend =>
+      evaluateExpressionWithSSA(extend.body, constantPropResult) match {
+        case b: Set[BitVecLiteral] if b.nonEmpty => applySingle(BitVectorEval.smt_zero_extend(extend.extension, _: BitVecLiteral), b)
+        case _                => Set()
+      }
+    case e: Extract =>
+      evaluateExpressionWithSSA(e.body, constantPropResult) match {
+        case b: Set[BitVecLiteral] if b.nonEmpty => applySingle(BitVectorEval.boogie_extract(e.end, e.start, _: BitVecLiteral), b)
+        case _               => Set()
+      }
+    case registerVariableWrapper: RegisterVariableWrapper =>
+      constantPropResult.collect({
+        case (k, v) if k == registerVariableWrapper => v
+      }).flatten.toSet
+      constantPropResult(registerVariableWrapper)
+    case variable: Variable =>
+      val registerVariableWrapper = RegisterVariableWrapper(variable)
+      constantPropResult.collect({
+        case (k, v) if k == registerVariableWrapper => v
+      }).flatten.toSet
```

Comment:
why does this need to use collect instead of just directly accessing the map? Is this a quirk of how the SSA is implemented - registerVariableWrapper equality is only checking if the variables share any SSA IDs? this defeats the whole purpose of a Map and is very unclear

Reply:
the result here isn't even used for anything? what does this do exactly? @yousifpatti

Reply:
I agree. An alternative lattice that does not use a map for constantPropResult needs to be used to avoid the problem of adding two identical registers which should work as a register tree structure. I will have to rewrite that part of the constantprop

Reply:
#193

Reply:
Part of the issue is that the result of this messy map access isn't even used, is that supposed to be the case

Reply:
That is true. The liner under it should have been deleted and the collected set should be the set being returned. I will fix that.

Reply:
Converted into a direct map access in the latest commit